pkg/config/agentmanagement.go

```diff
@@ -0,0 +1,162 @@
+	Enabled         bool             `yaml:"-"` // Derived from enable-features=agent-management
+	Url             string           `yaml:"api_url"`
+	BasicAuth       config.BasicAuth `yaml:"basic_auth"`
+	Protocol        string           `yaml:"protocol"`
+	PollingInterval string           `yaml:"polling_interval"`
+	CacheLocation   string           `yaml:"remote_config_cache_location"`
+
+	RemoteConfiguration RemoteConfiguration `yaml:"remote_configuration"`
+}
+
+// getRemoteConfig gets the remote config specified in the initial config, falling back to a local, cached copy
+// of the remote config if the request to the remote fails. If both fail, an empty config and an
+// error will be returned.
+func getRemoteConfig(expandEnvVars bool, initialConfig *Config, log *server.Logger) (*Config, error) {
+	if err := initialConfig.AgentManagement.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid initial config: %w", err)
+	}
+	remoteConfigBytes, err := fetchFromApi(initialConfig)
+	if err != nil {
+		level.Error(log).Log("msg", "could not fetch from API, falling back to cache", "err", err)
+		return getCachedRemoteConfig(initialConfig.AgentManagement.CacheLocation, expandEnvVars)
+	}
+	var remoteConfig Config
+
+	err = LoadBytes(remoteConfigBytes, expandEnvVars, &remoteConfig)
+	if err != nil {
+		level.Error(log).Log("msg", "could not load the response from the API, falling back to cache", "err", err)
+		return getCachedRemoteConfig(initialConfig.AgentManagement.CacheLocation, expandEnvVars)
+	}
+	level.Info(log).Log("msg", "fetched and loaded remote config from API")
+
+	if err = cacheRemoteConfig(initialConfig.AgentManagement.CacheLocation, remoteConfigBytes); err != nil {
+		level.Error(log).Log("err", fmt.Errorf("could not cache config locally: %w", err))
+	}
+	return &remoteConfig, nil
+}
+
+func getCachedRemoteConfig(cachePath string, expandEnvVars bool) (*Config, error) {
+	cachePath = filepath.Join(cachePath, cacheFilename)
+	var cachedConfig Config
+	if err := LoadFile(cachePath, expandEnvVars, &cachedConfig); err != nil {
+		return nil, fmt.Errorf("error trying to load cached remote config from file: %w", err)
+	}
+	return &cachedConfig, nil
+}
+
+func cacheRemoteConfig(cachePath string, remoteConfigBytes []byte) error {
+	cachePath = filepath.Join(cachePath, cacheFilename)
+	return os.WriteFile(cachePath, remoteConfigBytes, 0666)
+}
+
+// fetchFromApi fetches the raw bytes from the API based on the protocol specified in c.
+func fetchFromApi(c *Config) ([]byte, error) {
+	switch p := c.AgentManagement.Protocol; {
+	case p == "http":
+		return fetchConfig(c)
+	default:
+		return nil, fmt.Errorf("unsupported protocol for agent management api: %s", p)
+	}
+}
+
+// fetchConfig fetches the raw bytes of the config from the API specified in c.
+func fetchConfig(c *Config) ([]byte, error) {
+	httpClientConfig := &config.HTTPClientConfig{
+		BasicAuth: &c.AgentManagement.BasicAuth,
+	}
+
+	dir, err := os.Getwd()
```

Comment:
How does this behave on windows installed as a service?

Reply:
That's a good question (that I don't know the answer to, yet). This would only affect the case where the password file is being specified as a relative path ([relevant code where that dir would be used](https://github.com/prometheus/common/blob/main/config/config.go#L83)). An alternative would be to require that the password file be specified as an absolute path, at least for now?

Reply:
Well, I guess the answer would be: if the password file is specified with a relative path, it will look for the password file relative to the working directory of the windows service. (But, I don't know what that working directory would typically be)